### PR TITLE
Issue 1135: Associated constants

### PR DESCRIPTION
**Pull request by willmurphyscode**
This PR will make will use [associated constances](https://doc.rust-lang.org/1.6.0/book/associated-constants.html) instead of functions in 3 places:

- `diesel::query_builder::QueryId: const HAS_STATIC_QUERY_ID:` bool instead of `fn has_static_query_id() ->bool`
- `diesel::query_source::Column: const NAME: &'static str` instead of `fn name() -> &'static str`
- `diesel::types::FromSqlRow: const FIELDS_NEEDED: usize` (with default) instead of `fn fields_needed() -> usize { ... }`

These are the three places specifically mentioned in https://github.com/diesel-rs/diesel/issues/1135.

I have one stylistic question: Currently the `const WHATEVER` line replaces the first line of `fn whatever()`. If the line with `const WHATEVER` followed a line like `type SomeType`, I took out the blank line above. Otherwise, I left the blank lines as is. Is that how we want the style to be?

### CODE REVIEW

File: diesel/src/expression/unchecked_bind.rs [hunk 1]
```diff
@@ -32,10 +32,7 @@ impl<Query, Value, ST> QueryId for UncheckedBind<Query, Value, ST> where
     ST: QueryId,
 {
     type QueryId = UncheckedBind<Query::QueryId, (), ST::QueryId>;
-
-    fn has_static_query_id() -> bool {
-        Query::has_static_query_id() && ST::has_static_query_id()
-    }
+    const HAS_STATIC_QUERY_ID: bool = Query::HAS_STATIC_QUERY_ID && ST::HAS_STATIC_QUERY_ID;
```
Comment (killercup) [inline diesel/src/expression/unchecked_bind.rs]: Wow, that works? Nice!

File: diesel/src/macros/query_id.rs [hunk 2]
```diff
@@ -17,31 +14,22 @@ macro_rules! impl_query_id {
             $($ty_param: $crate::query_builder::QueryId),*
         {
             type QueryId = $name<$($ty_param::QueryId),*>;
-
-            fn has_static_query_id() -> bool {
-                $($ty_param::has_static_query_id() &&)* true
-            }
+            const HAS_STATIC_QUERY_ID: bool = $($ty_param::HAS_STATIC_QUERY_ID &&)* true;
```
Comment (killercup) [inline diesel/src/macros/query_id.rs]: Now that is just awesome

File: diesel/src/query_builder/query_id.rs [hunk 3]
```diff
@@ -3,10 +3,10 @@ use super::QueryFragment;
 
 pub trait QueryId {
     type QueryId: Any;
-    fn has_static_query_id() -> bool;
+    const HAS_STATIC_QUERY_ID: bool = true;
```
Comment (Eijebong) [inline diesel/src/query_builder/query_id.rs]: This one doesn't respect what you said about letting a blank line between each type of declaration

Comment (willmurphyscode) [reply diesel/src/query_builder/query_id.rs]: Thanks! I must've missed that one. I'll push up a change in a minute.

**Comment by willmurphyscode** [comment]
It looks like the `diesel_tests` project failed to build on Windows for Postgres. I'm a bit surprised, since I would expect this change either to succeed or to fail everything everywhere. 

I'm also surprised that this build: https://ci.appveyor.com/project/sgrif/diesel/build/1.0.1006 passed while this build: https://ci.appveyor.com/project/sgrif/diesel/build/1.0.1007 failed when the commit between them was this: https://github.com/diesel-rs/diesel/pull/1145/commits/c0bbc9f55263c95d514236c796e9c1965ab5f2f8 which has only a whitespace change. 

I'll attempt to repro this myself, but honestly my Windows computer is not set up for development at all, so I could use a little bit of assistance. (Also, it might be worth investigating a pipeline issue, since apparently a whitespace-only commit broke a build.)

**Comment by willmurphyscode** [comment]
It looks like https://github.com/diesel-rs/diesel/pull/1146 is failing CI builds with the same error, but also failing them for Postgres on Linux, not just Windows.

**Pull request merged**